### PR TITLE
EZP-25358: ezxmltext:convert-to-richtext command prototype

### PR DESCRIPTION
**Pull request by bdunogier**
> Status: partially tested prototype

Converted from procedural script provided by @pspanja, related doc being added to kernel: https://github.com/ezsystems/ezpublish-kernel/pull/1560

Will, at database level:
- replace `ezxmltext` field definitions with `ezrichtext` ones
- convert the content of `ezxmltext` fields to richtext

Regular output will display number of converted field definitions and values. Actual converted values will be logged, and can be displayed by adding the `-v` command option.

### Testing done
Tried on the `ezdemo` data set from eZ Publish Platform 5.4.0.

### TODO
- [x] Document
- [x] Use real DI (maybe)
- [x] @todo in the code about field def null value
- [x] Replace `var_dump` and `var_export` with proper output
- [x] Test in practice
- [x] Check database API usage (was using ezcDb)

### CODE REVIEW

File: bundle/Command/ConvertXmlTextToRichTextCommand.php [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        $convertedDocumentNormalized = new DOMDocument();
+        $convertedDocumentNormalized->loadXML($convertedDocument->saveXML());
+
+        $errors = $this->validator->validate($convertedDocument);
+
+        $result = $convertedDocumentNormalized->saveXML();
+
+        if (!empty($errors)) {
+            var_export($result);
+            var_export($errors);
```
Comment (andrerom) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: Additional todo: Change these and use of `var_dump` to use OutputInterface, which also contains progress features.

File: bundle/Command/ConvertXmlTextToRichTextCommand.php [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+                )
+            );
+            $updateQuery->prepare()->execute();
+        }
+
+
+        var_dump('converted field rows: ' . $count);
+    }
+
+    function convertFieldDefinitions()
```
Comment (andrerom) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: Nitpick, but since this is executed before `convertFields` it would be better if it was placed before in the class as well.

File: bundle/Command/ConvertXmlTextToRichTextCommand.php [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        while ($row = $statement->fetch(PDO::FETCH_ASSOC)) {
+            var_dump("===============", $row['id']);
+            if (empty($row['data_text'])) {
+                $inputValue = Value::EMPTY_VALUE;
+            } else {
+                $inputValue = $row['data_text'];
+            }
+
+            $converted = convert($inputValue);
```
Comment (andrerom) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: ` $converted = $this->convert($inputValue);`

File: bundle/Command/ConvertXmlTextToRichTextCommand.php [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        for ($i = 0; $i < $nodes->length; ++$i) {
+            $nodes->item($i)->parentNode->removeChild($nodes->item($i));
+        }
+    }
+
+    function convert($xmlString)
+    {
+        $inputDocument = $this->createDocument($xmlString);
+
+        removeComments($inputDocument);
```
Comment (andrerom) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: `$this-> removeComments($inputDocument);`

**Comment by andrerom** [comment]
We should update composer.json dep for this as well to require 6.x kernel, right now it is `    "ezsystems/ezpublish-kernel": "~5.4.5 | ~6.0@dev"`

File: bundle/Command/ConvertXmlTextToRichTextCommand.php [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+        $updateQuery = $this->db->createUpdateQuery();
+        $updateQuery->update($this->db->quoteIdentifier('ezcontentclass_attribute'));
+        $updateQuery->set(
+            $this->db->quoteIdentifier('data_type_string'),
+            $updateQuery->bindValue('ezrichtext', null, PDO::PARAM_STR)
+        );
+        $updateQuery->set(
+            $this->db->quoteIdentifier('data_text2'),
+            // TODO maybe null and not empty string
```
Comment (andrerom) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: This is tagPreset.
data_text2 is `\`data_text2\` varchar(50) DEFAULT NULL`, however legacy had:
```
                                         'data_text2' => array( 'name' => 'DataText2',
                                                                'datatype' => 'text',
                                                                'default' => '',
                                                                'required' => true ),
```

So it depends on if we want to fill database with legacy like value "", or if we want to use new stack like default value "null".

Comment (bdunogier) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: I really don't have a preference;..

Comment (andrerom) [inline bundle/Command/ConvertXmlTextToRichTextCommand.php]: I'd agree, but you'll need to test how it works setting it to null, or otherwise emptying column.

**Comment by andrerom** [comment]
+1 with null and doc in form you have already planned.

**Comment by bdunogier** [comment]
> with doc

Doc added, README.md and command description.

> remove of 5.4 in composer requirements.

afaict it is already removed in [master](https://github.com/ezsystems/ezplatform-xmltext-fieldtype/blob/master/composer.json#L13).

**Comment by bdunogier** [comment]
Last todo taken care of. I have re-executed the script with it, and displayed one of the converted pieces of text, looks okay. Needs to be re-tested though.
